build/changelog.py

```diff
@@ -25,6 +26,31 @@ def get_commit_message(commit_id):
         commit_id=commit_id)
     return run(cmd)
 
+def get_author_email(commit_id):
+    cmd = "git show -s --format=%ae {}".format(commit_id)
+    return run(cmd)
+
+def get_maintainers():
+    with open("../MAINTAINERS.md", "r") as f:
+        contents = f.read()
+    maintainers = re.findall(r"[^\s]+@[^\s]+", contents)
+    return maintainers
+
+def get_commit_author(commit_id):
+    url = "https://api.github.com/repos/open-policy-agent/opa/commits/{}".format(commit_id)
+    author_email = get_author_email(commit_id).replace("\n", "")
+    
+    if author_email in maintainers:
+        return author_email[:author_email.index("@")]
```

Comment:
I'd prefer if we didn't have to do extra `git` shell commands and use the emails to match things up (in my experience git commit emails are kind of brittle, eg I have `east.patrick@gmail.com` for my github account and in the maintainers doc, but on occasion use a different tool that might tag a commit with my work email `patrick@styra.com`.

I was thinking we could cheat and use the github account names from `MAINTAINERS.md` but parsing that out seems like kind of a headache compared to just making one more URL request at the start and caching the set of maintainer usernames.

I think the best thing we can do is to use the list from https://api.github.com/orgs/open-policy-agent/members and always query for the account name via the github commit API.

You mentioned that there were problems with rate limiting for people that are not signed in, I think that is OK. We can just add an option to supply an API token for the script. We'll just need to add in a header like `"Authorization: token ${GITHUB_API_TOKEN}"` to each request.

This should be pretty easy for anyone to do that is creating the release.

Note that it will need `read:org` scope to be able to list the full group of OPA members (not everyone is a public member).

If an API token is not supplied the script could maybe log a warning or something, but should just continue trying to generate the changelog. If there are too many commits it might hit the ratelimit, and/or miss private group members.. but that isn't a big deal IMO.

@tsandall any objections to adding the requirement of providing an API token with `make release-patch`?

Reply:
Got it! 

I've made the requested changes, the only thing that might be needed to be changed now is how the token is accepted. I've added a cli flag for it but I can change that to use os.env (or any other method) if required. 
Please have a look at it and let me know if anything needs to be changed.

Thanks!

Reply:
I chatted about this offline with @patrick-east and we had a few thoughts:

1. Including @ mentions in the change log should be "best effort". We don't want to make the release process more onerous in order to ensure 100% coverage. E.g., interrupting the release process because we couldn't fetch the GitHub username of someone who fixed a typo inside of a comment doesn't feel like the right approach.

2. Requiring a GitHub token to run the script is fine but it'll be a bit annoying (now you have you to go into GitHub and generate a token or save it somewhere or probably just rely on shell history.) If we can avoid requiring a token that would be nice.

3. I looked at the numbers and over the past 4 releases (+between now and the v0.19) we've averaged 13.8 unique committers per release. The majority of the commits come from maintainers and there's the standard long tail of smaller patches you get in most open source projects. If we assume that a given release has 10-20 external contributors for the time being that means we could only have to make 10-20 GitHub API calls which feels like it ought to be well within the ratelimit. Also, some external contributors will supply multiple patches so my guess is it's closer to 10.

Given the above, here's what we should do:

* Parse the signed-off-by lines of the commit messages. This gives us a name and email. A quick look at the log for v0.17..v0.18 (which had the most commits) shows that all the sign-offs look reasonable.

* Exclude commits from emails that show up in the MAINTAINERS.md file. A simple string contains check would probably work fine.

* For each commit from from the remaining contributors fetch the username from GitHub. Cache the result and key it off the name/email from the signed-off-by line.

If we do this we should invoke the GitHub API as little as possible be able to do it within the ratelimit. If we're still hitting ratelimits with this approach I'd be surprised but perhaps we could look at whether GitHub supports a bulk GET that allows you to fetch info for multiple commits or a range of commits where the result contains the login.

Reply:
Thanks for such a detailed response @tsandall!

> 1. Including @ mentions in the change log should be "best effort". We don't want to make the release process more onerous in order to ensure 100% coverage. E.g., interrupting the release process because we couldn't fetch the GitHub username of someone who fixed a typo inside of a comment doesn't feel like the right approach.
> 2. Requiring a GitHub token to run the script is fine but it'll be a bit annoying (now you have you to go into GitHub and generate a token or save it somewhere or probably just rely on shell history.) If we can avoid requiring a token that would be nice.

Both of the commits that I have made work well even if you do not provide the token.
1. The first one parses the emails of the maintainers from the MAINTAINERS.md file and uses a git shell command to get the author email from the git log. It then makes an API request to get the GitHub id only if the email is not present in the .md file, thereby reducing the number of calls (and thus remaining under the rate limit). The main issue with the approach was that it needed to run an extra shell command and using email is a bit flaky.

2. The second method makes an API call to cache the maintainers github id's and then for each commit gets the github id of the author of the commit (and appends the entry authored by @<user> if required). In this case, if you don't pass the token, the script will just skip adding the mention and the output of the changelog file generated will be the same as before (no mentions), since the rate limit will be encountered sooner, and most of the commits are from the maintainers so while you are making the api calls for them, you won't be adding them to the changelog. [This](https://pastebin.com/Ceu9y4T0) is the output I get when I run it on the current branch. (Also, the problem here is that if, for any reason, you have to run the script again (you interrupt the program or something), you would have already hit the rate limit so the output generated will be no different than if this feature was not there)

> 3. I looked at the numbers and over the past 4 releases (+between now and the v0.19) we've averaged 13.8 unique committers per release. The majority of the commits come from maintainers and there's the standard long tail of smaller patches you get in most open source projects. If we assume that a given release has 10-20 external contributors for the time being that means we could only have to make 10-20 GitHub API calls which feels like it ought to be well within the ratelimit. Also, some external contributors will supply multiple patches so my guess is it's closer to 10.
> 
> Given the above, here's what we should do:
> 
> * Parse the signed-off-by lines of the commit messages. This gives us a name and email. A quick look at the log for v0.17..v0.18 (which had the most commits) shows that all the sign-offs look reasonable.

That would require a shell command, right? 
I had used `git show -s --format=%ae commit-id`, is there a better way to get the email?

> * Exclude commits from emails that show up in the MAINTAINERS.md file. A simple string contains check would probably work fine.

Do you mean to read the contents of the file in a string and just check `if email in file`? It feels a bit weird to me and I would rather go with the regex match to store the maintainers in a list.

> * For each commit from the remaining contributors fetch the username from GitHub. Cache the result and key it off the name/email from the signed-off-by line.

> If we do this we should invoke the GitHub API as little as possible be able to do it within the ratelimit. If we're still hitting ratelimits with this approach I'd be surprised but perhaps we could look at whether GitHub supports a bulk GET that allows you to fetch info for multiple commits or a range of commits where the result contains the login.

This sounds very similar to the previous commit (apart from the third point) to me, wdyt?

/cc @patrick-east